File: sections/40_data_structures_mixed_contest/problems/d_salary_queries/solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct Fenwick {
    vector<int> bit;
    explicit Fenwick(int n) : bit(n + 1, 0) {}

    void add(int index, int delta) {
        int size = static_cast<int>(bit.size());
        for (; index < size; index += index & -index) {
            bit[index] += delta;
        }
    }

    int sum(int index) const {
        int result = 0;
        for (; index > 0; index -= index & -index)
            result += bit[index];
        return result;
    }
};

struct Operation {
    char type;
    int first;
    long long second;
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, q;
    if (!(cin >> n >> q)) return 0;
    vector<long long> salary(n + 1), coords;
    for (int i = 1; i <= n; i++) {
        cin >> salary[i];
        coords.push_back(salary[i]);
    }

    vector<Operation> operations;
    operations.reserve(q);
    for (int i = 0; i < q; i++) {
        char type;
        long long first, second;
        cin >> type >> first >> second;
        operations.push_back({type, (int)first, second});
        if (type == '!') coords.push_back(second);
    }

    sort(coords.begin(), coords.end());
    auto unique_end = unique(coords.begin(), coords.end());
    coords.erase(unique_end, coords.end());
    auto rank_of = [&](long long value) {
        auto iterator = lower_bound(
            coords.begin(), coords.end(), value
        );
        return int(iterator - coords.begin()) + 1;
    };

    Fenwick bit((int)coords.size());
    for (int i = 1; i <= n; i++) bit.add(rank_of(salary[i]), 1);

    for (const Operation& operation : operations) {
        if (operation.type == '!') {
            int employee = operation.first;
            bit.add(rank_of(salary[employee]), -1);
            salary[employee] = operation.second;
            bit.add(rank_of(salary[employee]), 1);
        } else {
            long long low = operation.first;
            long long high = operation.second;
            auto low_iterator = lower_bound(
                coords.begin(), coords.end(), low
            );
            auto high_iterator = upper_bound(
                coords.begin(), coords.end(), high
            );
            int before_low = int(low_iterator - coords.begin());
            int through_high = int(high_iterator - coords.begin());
            int answer = (
                bit.sum(through_high) - bit.sum(before_low)
            );
            cout << answer << '\n';
        }
    }
    return 0;
}

File: sections/40_data_structures_mixed_contest/problems/d_salary_queries/solution.py
import bisect
import sys

class Fenwick:

    def __init__(self, size: int) -> None:
        self.bit = [0] * (size + 1)

    def add(self, index: int, delta: int) -> None:
        while index < len(self.bit):
            self.bit[index] += delta
            index += index & -index

    def prefix_sum(self, index: int) -> int:
        total = 0
        while index:
            total += self.bit[index]
            index -= index & -index
        return total

def main() -> None:
    tokens = sys.stdin.buffer.read().split()
    if not tokens:
        return
    pointer = 0
    n = int(tokens[pointer])
    q = int(tokens[pointer + 1])
    pointer += 2
    initial = tokens[pointer:pointer + n]
    salary = [0] + [int(value) for value in initial]
    pointer += n
    operations: list[tuple[bytes, int, int]] = []
    coords = salary[1:]
    for _ in range(q):
        kind = tokens[pointer]
        first = int(tokens[pointer + 1])
        second = int(tokens[pointer + 2])
        pointer += 3
        operations.append((kind, first, second))
        if kind == b'!':
            coords.append(second)
    coords = sorted(set(coords))
    bit = Fenwick(len(coords))

    def rank_of(value: int) -> int:
        return bisect.bisect_left(coords, value) + 1
    for value in salary[1:]:
        bit.add(rank_of(value), 1)
    output: list[str] = []
    for kind, first, second in operations:
        if kind == b'!':
            employee, new_salary = (first, second)
            bit.add(rank_of(salary[employee]), -1)
            salary[employee] = new_salary
            bit.add(rank_of(new_salary), 1)
        else:
            low, high = (first, second)
            before_low = bisect.bisect_left(coords, low)
            through_high = bisect.bisect_right(coords, high)
            answer = (
                bit.prefix_sum(through_high)
                - bit.prefix_sum(before_low)
            )
            output.append(str(answer))
    print('\n'.join(output))
if __name__ == '__main__':
    main()
